rain_api_core/bucket_map.py

```diff
@@ -168,3 +224,172 @@ def _parse_access_control(bucket_map: dict) -> dict:
+        return _is_accessible(required_groups, self.groups)
+
+    def generate_policy(self, entries: Iterable[BucketMapEntry]) -> dict:
+        full_access_entries = []
+        partial_access_entries = []
+
+        for entry in entries:
+            if self._is_whole_bucket_accessible(entry):
+                collection = full_access_entries
+            else:
+                collection = partial_access_entries
+            collection.append(entry)
+
+        full_access_statement = ({
+            "Effect": "Allow",
+            "Action": ["s3:GetObject", "s3:ListBucket"],
+            "Resource": [
+                resource
+                for entry in full_access_entries
+                for resource in (
+                    f"arn:aws:s3:::{entry.bucket}",
+                    f"arn:aws:s3:::{entry.bucket}/*"
+                )
+            ]
+        },) if full_access_entries else ()
+
+        statement = [
+            *full_access_statement,
+            *(
+                statement
+                for entry in partial_access_entries
+                for statement in self._generate_iam_statements(entry)
+            )
+        ]
+        policy = {
+            "Version": "2012-10-17",
+            "Statement": statement or [
+                # Special case noop statement that will never match anything.
+                # We need this because IAM doesn't allow empty statement lists
+                {
+                    "Effect": "Allow",
+                    "Action": ["s3:GetObject", "s3:ListBucket"],
+                    "Resource": ["*"],
+                    "Condition": {
+                        "StringNotLike": {
+                            "s3:prefix": [""]
+                        }
+                    }
+                }
+            ]
+        }
+
+        return policy
+
+    def _is_whole_bucket_accessible(self, entry: BucketMapEntry) -> bool:
+        if entry._access_control is None:
+            return True
+
+        return all(
+            self._is_accessible(required_groups)
+            for required_groups in entry._access_control.values()
+        )
+
+    def _generate_iam_statements(self, entry: BucketMapEntry) -> Generator[dict, None, None]:
+        assert entry._access_control, "Public buckets should be handled already"
+
+        for condition in self._generate_iam_conditions(entry._access_control):
+            statement = {
+                "Effect": "Allow",
+                "Action": ["s3:GetObject", "s3:ListBucket"],
+                "Resource": [
+                    f"arn:aws:s3:::{entry.bucket}",
+                    f"arn:aws:s3:::{entry.bucket}/*"
+                ]
+            }
+            if condition:
+                statement["Condition"] = condition
+
+            yield statement
+
+    def _generate_iam_conditions(self, access_control: dict) -> Generator[dict, None, None]:
+        access_control = dict(access_control)
+        access_control.setdefault("", None)
```

Comment:
Maybe the empty key should be guaranteed to be part of the `access_control` dict. That would also simplify the logic in `get_required_groups`